### ISSUE TITLE
Provide CosmosDB native repository

### ISSUE THREAD

**Issue by blackeyedyl** [Azure/azure-mobile-apps#785]
**Is your feature request related to a problem? Please describe.**
In our project we're using Comsos DB as our backend database and we are using it with both single and hierarchical/hashed partition keys. The current implementation using EF to access the Cosmos DB is limited to partition keys that are the same as the ID and the EF Core team does not yet support hierarchical/hashed partition keys.

This means that we can't use the provided `TableController` because to get or delete an item we need the client to send the partition key in order to do the lookup. So, long story short, we had to roll our own controllers. This works fine for successful http requests, but if we want to return a `PreconditionFailedException` then the client will never get the payload of the exception because the `DatasyncExceptionsAttribute` is `internal`.

**Describe the solution you'd like**
For now, I think the simplest solution would be to change all of the attributes found on the default `TableController` to be changed from `internal` to `public`:
- DatasyncExceptionsAttribute
- DatasyncProtocolVersionFilterAttribute
- AzureMobileBackwardsCompatibilityFilter

As well as the following helpers that are used in the `TableController`:
- HttpRequestExtensions
- ITableDataExtensions

**Describe alternatives you've considered**
Another option would be to make a Cosmos DB specific library that would allow the client to pass in the required partition key information, but I believe these classes would still need to be made public to do this.

**Additional context**
I've created a PR for this: #786

**Comment by blackeyedyl** [comment]
Thanks for the quick response, I would love some input on this. Note that we have implemented our own repositories, but as you will see below, they have a similar issue as the TableController.

The issue is that in Cosmos DB, both the ID and partition key are required for unique identification of a document. Your current ReadAsync() and DeleteAsync() methods in TableController<TEntity> and IRepository<TEntity> only accept the ID. This is problematic, especially when using [hierarchical/hashed partition key](https://learn.microsoft.com/en-us/azure/cosmos-db/hierarchical-partition-keys?tabs=net-v3%2Cbicep#choose-your-hierarchical-partition-keys), as you may need additional properties to perform reads/deletes.

![image](https://github.com/Azure/azure-mobile-apps/assets/6576706/920391a7-b302-4561-9347-404f80f27b8a)

![image](https://github.com/Azure/azure-mobile-apps/assets/6576706/3c0eed3d-e460-4490-bcbb-3abf0c19186c)

Using a plain query to select by only the ID incurs extra costs. A point read operation with both ID and partition key costs 1 Request Unit (RU), but a query based only on the ID will have variable cost depending on the number of partitions and object size.

**Comment by blackeyedyl** [comment]
Funny, I was just thinking about that as an option.

**Comment by adrianhall** [comment]
Let's re-title this issue (and update the description) to "Provide CosmosDB native repository"

**Comment by blackeyedyl** [comment]
I'll see how things go with using a compound id with our implementation and maybe I'll be able to create a comsos sdk project if I'm successful. Thanks 😊

**Comment by blackeyedyl** [comment]
@adrianhall FYI I've started this on a branch on my repo here: https://github.com/Azure/azure-mobile-apps/compare/main...blackeyedyl:azure-mobile-apps:cosmos-repository

The easy bits are in, next is to figure out how to handle the partition keys (I've added `// TODO` everywhere we need a partion key.

I've reused the EtagEntityTableData from the EFCore implementation. Would probably be good to re-use that instead of copy pasting but wasn't sure where best to put that along with the tests for it.

**Comment by adrianhall** [comment]
I would suggest a new project (let's say Microsoft.AspNetCore.Cosmos) with a CosmosRepository and a CosmosTableData - just like the LiteDB and EFCore versions.  This way, the repo stands on its own and doesn't require entity framework as a dependency.

**Comment by adrianhall** [comment]
Looks like you have opted for CosmosTableData - which is a good plan.

**Comment by blackeyedyl** [comment]
@adrianhall looks like Cosmos SDK doesn't support LongCount():

> Microsoft.Azure.Cosmos.Linq.DocumentQueryException: Method 'LongCount' is not supported.

https://github.com/Azure/azure-cosmos-dotnet-v3/issues/1505
https://github.com/Azure/azure-cosmos-dotnet-v3/issues/738

**Comment by blackeyedyl** [comment]
> Looks like you have opted for CosmosTableData - which is a good plan.

Just building based on the other implementations. My only concern with this approach is that the model must inherit from CosmosTableData which works if you have a client model and a server model. I would like to support a model that's shared between the client and server. Thoughts?

**Comment by adrianhall** [comment]
The short version is "no - don't do that".  There is generally private data that is used in determining if the user should have access to the data on the server side that isn't (and can't be) present on the client side.  There may be other restrictions as well - data types are not necessarily common, for instance.

**Comment by adrianhall** [comment]
For the LongCount() problem, I was thinking of adding a CountAsync(queryable) to the repository anyhow - this might be the appropriate time to do that functionality.

**Comment by adrianhall** [comment]
Ref: PR #788 - once that is merged, we won't be using .LongCount() any more.

**Comment by blackeyedyl** [comment]
Running into another OData issue which seems a bit trickier to fix. When I run the sample app I added in my branch and try to `$select` a column I get the following:

> Newtonsoft.Json.JsonSerializationException: Self referencing loop detected for property 'declaringType' with type 'Microsoft.OData.Edm.EdmEntityType'. Path 'typedProperty.schemaElements[0].declaredKey[0]'.

This seems to be a "known" issue with OData and Cosmos, but it's unclear on how to fix it from my preliminary search:

Some folks say that adding `AddODataNewtonsoftJson()` fixes it, but for me it has no effect:
https://github.com/JamesNK/Newtonsoft.Json/issues/2649#issuecomment-1169616961

When I tried to add `ReferenceLoopHandling = ReferenceLoopHandling.Ignore` I end up with an infinite loop similar to the issue reported here:
https://github.com/OData/AspNetCoreOData/issues/87#issuecomment-960377856

**Comment by adrianhall** [comment]
I'm surprised the EF Core driver doesn't have this problem.  I'll dig a little bit this weekend (now I've got 6.0.4 out)

**Comment by blackeyedyl** [comment]
Hi @adrianhall, have you had any time to think on that serialization issue?

The cosmos server repository is making good progress, and the code in my branch is functioning well. However, the datasync client code to check for a valid ID is causing some challenges with the "constructed key" concept, here is the relevant code:

https://github.com/Azure/azure-mobile-apps/blob/940cfa18de2e0ec270b8828e95938f9f022c0f1c/sdk/dotnet/src/Microsoft.Datasync.Client/Utils/Arguments.cs#L25

https://github.com/Azure/azure-mobile-apps/blob/940cfa18de2e0ec270b8828e95938f9f022c0f1c/sdk/dotnet/src/Microsoft.Datasync.Client/Utils/Arguments.cs#L121

Partition keys can be any non-null string up to 100 characters in length, and there can be up to three partition keys. Would it make sense to allow a client option to override the default ID regex?

**Comment by adrianhall** [comment]
Your ID must conform to a path segment definition in RFC 2396 and fit into the ID field of the table.   It's currently between 1 and 126 characters.  Is 126 not big enough?  I get that partition keys CAN be that length, but can we put some reasonable restrictions on them?

**Comment by adrianhall** [comment]
... and, no, I've not had time to download your code and prod yet.  I'm setting aside Friday for this task.

**Comment by blackeyedyl** [comment]
> Your ID must conform to a path segment definition in RFC 2396 and fit into the ID field of the table. It's currently between 1 and 126 characters. Is 126 not big enough? I get that partition keys CAN be that length, but can we put some reasonable restrictions on them?

The 126 length is likely fine, its more about the allowed characters. I can work around the specific issue I'm encountering.

**Comment by adrianhall** [comment]
I'm happy to add additional characters, as long as they aren't "wildcard" characters.  Preferably, only characters that don't need to be encoded on the URL (hence the RFC 2396).  If they have to be URL encoded, that can cause other issues.

Want to put in a PR with your preferred character set?

**Comment by blackeyedyl** [comment]
Thanks @adrianhall, but the specific character would need to be encoded, so no worries. The restrictions makes sense.

Now I'm struggling with where and when to do the translation of the id so that things work with the SQLite implementation of OfflineTable which uses JObjects. I'm not sure if I'm running into a technical limitation or I'm just going about things incorrectly. 

Quoting the original idea:
> The ID in the DTO is up to you - it's an opaque string. There is no reason you can't construct an ID in the DTO from the ID+PartitionKey in the Cosmos table.
>
> In your repository, have an ID translator (both ways) and use it throughout as necessary.

Given the following TodoItems partitioned by the UserId property:
- **id /UserId**
- 1 /Bilbo
- 2 /Bilbo
- 1 /Samwise 
- 3 /Bilbo
- 2 /Samwise 
> (Cosmos DB allows ID re-use on separate partitions in reality we are generally using guids)

Since the OfflineTable seems to save data from the get API calls to the SQLite database from the JObject so it's up to the API to return a constructed ID for the TodoItems (1:Bilbo, 2:Samwise , etc), and likewise the API expects any data objects sent to the server to have an ID set to the constructed ID. 

I'm unsure on how to structure the server DTO objects so that Repository.Query() serializes the objects with the constructed ID and also use those same objects within the repository to save the ID to the database without the ":UserId" in the ID.

I hope my explanation is clear. I tried to avoid making it too lengthy but could really use some direction here.

FYI I've updated the new cosmos sample in my branch to have a container partitioned on the UserId to help with this discussion: 
https://github.com/blackeyedyl/azure-mobile-apps/tree/779b15b4515e2ad7bcbf6150a2c2a46e16186cd7/samples/Samples.Cosmos.SDK

**Comment by adrianhall** [comment]
I downloaded the branch to see what was going on, but there were too many compile time errors for the sample.  I'll take a stronger look tomorrow based on just the SDK and unit tests.

**Comment by blackeyedyl** [comment]
Compile time errors? That's embarrassing! I'll double check the branch this evening.

FYI I'm using VS 2022 on Windows

**Comment by blackeyedyl** [comment]
Looks like I may have been missing a few commits, I cloned and compiled the sample and SDK on my Mac and it seems to be running for me.

**Comment by blackeyedyl** [comment]
I have found a "solution", but I don't know if I am happy with it as it has a house of cards feel to it. Because Cosmos and the Controller are using the same model object I added some extra code to the CosmosTableData object to allow the ID to be translated and return the constructed ID when the AsQueryable() is called. This works great, except when creating and replacing the object I need to pass in an entity with a non-constructed ID. For now I have worked around this by adding a trigger to my container that fixes the ID before the record is upserted - which is the house of cards part.

I've checked in the sample, updates and tests but like I said this feels like a workaround rather than a solution because everything needs to be setup just right for it to work.

Note: I also added a `.http` file to the sample to help with testing.

**Comment by blackeyedyl** [comment]
I removed my previous comment because the trigger solution wasn't viable. I think I have come up with something that might work:

1.  CosmosTableData object can be extended to return a constructed ID, which is needed so that calls to AsQueryable() returns the constructed ID.
2. Parse the ID and PartitionKey from the contructed ID when calling ReadAsync()
3. When updating the database with ReplaceAsync() and CreateAsync() I'm using the Cosmos serializer to convert the passed in entity to a JObject update it with the database ID. This JObject is then used to update the database.

More tests and testing is needed this week, but it's feeling like this could work if we always want to return a constructed ID.

**Comment by blackeyedyl** [comment]
Hi @adrianhall I think the native cosmos library is at a point where feedback would be appreciated. We are using it in-house, and it seems to be working quite well. We are using both the remote and offline tables on the client side, and a mix of single and hierarchical partition keys on the Cosmos DB side. 

The main challenges were:
- OData select is not supported which I believe this can be a known limitation - thoughts?
- Handling of the constructed IDs: the DTO model object needed to include code to construct and deconstruct the IDs to work with the standard CRUD operations as well as the AsQueryable() operation. I also needed to edit the cosmos JSON stream to ensure the deconstructed ID was saved to the cosmos database.
- Handling of dates: cosmos doesn't have a date field, so I had to ensure that the dates/offsets were saved to the cosmos containers with the correct format. I add a default cosmos serializer class to ensure dates were saved to the database in ISO 8601 format instead of the interchange format used by the sync library.

The only "TODO" tasks I have remaining is to add better handling of double and bool partition key values. If you believe that this new project is ready to be merged into your repository, I'll start working on this if it's deemed important.

My 'cosmos-repository' branch can be found here: 'https://github.com/blackeyedyl/azure-mobile-apps/tree/cosmos-repository' - I can create a PR if that would be preferred.

The last time you checked out the project, you mentioned some compile errors. Hopefully, those have been resolved, but if not, please let me know the specific errors so that I can investigate. It's possible that I might have used some newer C# language features that need adjustment.

**Comment by adrianhall** [comment]
I think OData select is a big deal.  It breaks the promise that the client doesn’t need to know how the server installed.  The rest of it is fairly straight forward and looks good.  I’ll talk to the Cosmos team to see what we can do about the select problem next week.

**Comment by blackeyedyl** [comment]
I tried looking into the EF Core implementation of CosmosDB, and at a glance they might be handling the "select" operation on the client-side. Would it be a good idea to enable the repository to extend when client-side evaluation should take place by adding a repository call as part of the `IsClientSideEvaluationException()` logic. Alternatively, the repository could inform the table controller about which OData capabilities should be handled client-side. 

```
/// <summary>
/// Determines if a particular exception indicates a client-side evaluation is required.
/// </summary>
/// <param name="ex">The exception that was thrown by the service-side evaluator</param>
/// <returns>true if a client-side evaluation is required.</returns>
[NonAction]
internal static bool IsClientSideEvaluationException(Exception ex)
    => ex != null && (ex is InvalidOperationException || ex is NotSupportedException);
```

Based on the comments in the Query() method in the TableController this is needed for EF Core for similar reasons. 

Below is the full exception:
```
System.Reflection.TargetInvocationException: Exception has been thrown by the target of an invocation.
 ---> Newtonsoft.Json.JsonSerializationException: Self referencing loop detected for property 'declaringType' with type 'Microsoft.OData.Edm.EdmEntityType'. Path 'typedProperty.schemaElements[0].declaredKey[0]'.
   at Newtonsoft.Json.Serialization.JsonSerializerInternalWriter.CheckForCircularReference(JsonWriter writer, Object value, JsonProperty property, JsonContract contract, JsonContainerContract containerContract, JsonProperty containerProperty)
   at Newtonsoft.Json.Serialization.JsonSerializerInternalWriter.CalculatePropertyValues(JsonWriter writer, Object value, JsonContainerContract contract, JsonProperty member, JsonProperty property, JsonContract& memberContract, Object& memberValue)
   at Newtonsoft.Json.Serialization.JsonSerializerInternalWriter.SerializeObject(JsonWriter writer, Object value, JsonObjectContract contract, JsonProperty member, JsonContainerContract collectionContract, JsonProperty containerProperty)
   at Newtonsoft.Json.Serialization.JsonSerializerInternalWriter.SerializeList(JsonWriter writer, IEnumerable values, JsonArrayContract contract, JsonProperty member, JsonContainerContract collectionContract, JsonProperty containerProperty)
   at Newtonsoft.Json.Serialization.JsonSerializerInternalWriter.SerializeObject(JsonWriter writer, Object value, JsonObjectContract contract, JsonProperty member, JsonContainerContract collectionContract, JsonProperty containerProperty)
   at Newtonsoft.Json.Serialization.JsonSerializerInternalWriter.SerializeList(JsonWriter writer, IEnumerable values, JsonArrayContract contract, JsonProperty member, JsonContainerContract collectionContract, JsonProperty containerProperty)
   at Newtonsoft.Json.Serialization.JsonSerializerInternalWriter.SerializeObject(JsonWriter writer, Object value, JsonObjectContract contract, JsonProperty member, JsonContainerContract collectionContract, JsonProperty containerProperty)
   at Newtonsoft.Json.Serialization.JsonSerializerInternalWriter.SerializeObject(JsonWriter writer, Object value, JsonObjectContract contract, JsonProperty member, JsonContainerContract collectionContract, JsonProperty containerProperty)
   at Newtonsoft.Json.Serialization.JsonSerializerInternalWriter.Serialize(JsonWriter jsonWriter, Object value, Type objectType)
   at Newtonsoft.Json.JsonSerializer.SerializeInternal(JsonWriter jsonWriter, Object value, Type objectType)
   at Newtonsoft.Json.JsonConvert.SerializeObjectInternal(Object value, Type type, JsonSerializer jsonSerializer)
   at Microsoft.Azure.Cosmos.Linq.ExpressionToSql.VisitConstant(ConstantExpression inputExpression, TranslationContext context)
   at Microsoft.Azure.Cosmos.Linq.ExpressionToSql.VisitNonSubqueryScalarExpression(Expression inputExpression, TranslationContext context)
   at Microsoft.Azure.Cosmos.Linq.ExpressionToSql.VisitNonSubqueryScalarExpression(Expression expression, ReadOnlyCollection`1 parameters, TranslationContext context)
   at Microsoft.Azure.Cosmos.Linq.ExpressionToSql.VisitScalarExpression(Expression expression, ReadOnlyCollection`1 parameters, TranslationContext context)
   at Microsoft.Azure.Cosmos.Linq.ExpressionToSql.VisitScalarExpression(Expression expression, TranslationContext context)
   at Microsoft.Azure.Cosmos.Linq.ExpressionToSql.VisitMemberAccess(MemberExpression inputExpression, TranslationContext context)
   at Microsoft.Azure.Cosmos.Linq.ExpressionToSql.VisitNonSubqueryScalarExpression(Expression inputExpression, TranslationContext context)
   at Microsoft.Azure.Cosmos.Linq.ExpressionToSql.VisitNonSubqueryScalarExpression(Expression expression, ReadOnlyCollection`1 parameters, TranslationContext context)
   at Microsoft.Azure.Cosmos.Linq.ExpressionToSql.VisitScalarExpression(Expression expression, ReadOnlyCollection`1 parameters, TranslationContext context)
   at Microsoft.Azure.Cosmos.Linq.ExpressionToSql.VisitScalarExpression(Expression expression, TranslationContext context)
   at Microsoft.Azure.Cosmos.Linq.ExpressionToSql.VisitMemberAssignment(MemberAssignment inputExpression, TranslationContext context)
   at Microsoft.Azure.Cosmos.Linq.ExpressionToSql.VisitBinding(MemberBinding binding, TranslationContext context)
   at Microsoft.Azure.Cosmos.Linq.ExpressionToSql.VisitBindingList(ReadOnlyCollection`1 inputExpressionList, TranslationContext context)
   at Microsoft.Azure.Cosmos.Linq.ExpressionToSql.VisitMemberInit(MemberInitExpression inputExpression, TranslationContext context)
   at Microsoft.Azure.Cosmos.Linq.ExpressionToSql.VisitNonSubqueryScalarExpression(Expression inputExpression, TranslationContext context)
   at Microsoft.Azure.Cosmos.Linq.ExpressionToSql.VisitNonSubqueryScalarExpression(Expression expression, ReadOnlyCollection`1 parameters, TranslationContext context)
   at Microsoft.Azure.Cosmos.Linq.ExpressionToSql.VisitScalarExpression(Expression expression, ReadOnlyCollection`1 parameters, TranslationContext context)
   at Microsoft.Azure.Cosmos.Linq.ExpressionToSql.VisitScalarExpression(LambdaExpression lambda, TranslationContext context)
   at Microsoft.Azure.Cosmos.Linq.ExpressionToSql.VisitSelect(ReadOnlyCollection`1 arguments, TranslationContext context)
   at Microsoft.Azure.Cosmos.Linq.ExpressionToSql.VisitMethodCall(MethodCallExpression inputExpression, TranslationContext context)
   at Microsoft.Azure.Cosmos.Linq.ExpressionToSql.Translate(Expression inputExpression, TranslationContext context)
   at Microsoft.Azure.Cosmos.Linq.ExpressionToSql.VisitMethodCall(MethodCallExpression inputExpression, TranslationContext context)
   at Microsoft.Azure.Cosmos.Linq.ExpressionToSql.Translate(Expression inputExpression, TranslationContext context)
   at Microsoft.Azure.Cosmos.Linq.ExpressionToSql.TranslateQuery(Expression inputExpression, IDictionary`2 parameters, CosmosLinqSerializerOptions linqSerializerOptions)
   at Microsoft.Azure.Cosmos.Linq.SqlTranslator.TranslateQuery(Expression inputExpression, CosmosLinqSerializerOptions linqSerializerOptions, IDictionary`2 parameters)
   at Microsoft.Azure.Cosmos.Linq.DocumentQueryEvaluator.HandleMethodCallExpression(MethodCallExpression expression, IDictionary`2 parameters, CosmosLinqSerializerOptions linqSerializerOptions)
   at Microsoft.Azure.Cosmos.Linq.DocumentQueryEvaluator.Evaluate(Expression expression, CosmosLinqSerializerOptions linqSerializerOptions, IDictionary`2 parameters)
   at Microsoft.Azure.Cosmos.Linq.CosmosLinqQuery`1.CreateFeedIterator(Boolean isContinuationExpected)
   at Microsoft.Azure.Cosmos.Linq.CosmosLinqQuery`1.GetEnumerator()+MoveNext()
   at System.Collections.Generic.List`1..ctor(IEnumerable`1 collection)
   at Microsoft.AspNetCore.OData.Query.Container.TruncatedCollection`1..ctor(IQueryable`1 source, Int32 pageSize, Boolean parameterize)
   at Microsoft.AspNetCore.OData.Query.ODataQueryOptions.LimitResults[T](IQueryable`1 queryable, Int32 limit, Boolean parameterize, Boolean& resultsLimited)
   at System.RuntimeMethodHandle.InvokeMethod(Object target, Void** arguments, Signature sig, Boolean isConstructor)
   at System.Reflection.MethodInvoker.Invoke(Object obj, IntPtr* args, BindingFlags invokeAttr)
   --- End of inner exception stack trace ---
   at Microsoft.AspNetCore.Datasync.TableController`1.CatchClientSideEvaluationException(Exception ex, String reason, Action clientSideEvaluator) in C:\Users\Dylan.Mills\source\repos\datasync\sdk\dotnet\src\Microsoft.AspNetCore.Datasync\Tables\TableController.cs:line 144
   at Microsoft.AspNetCore.Datasync.TableController`1.QueryAsync(CancellationToken token) in C:\Users\Dylan.Mills\source\repos\datasync\sdk\dotnet\src\Microsoft.AspNetCore.Datasync\Tables\TableController.cs:line 394
   at Microsoft.AspNetCore.Mvc.Infrastructure.ActionMethodExecutor.TaskOfIActionResultExecutor.Execute(ActionContext actionContext, IActionResultTypeMapper mapper, ObjectMethodExecutor executor, Object controller, Object[] arguments)
   at Microsoft.AspNetCore.Mvc.Infrastructure.ControllerActionInvoker.<InvokeActionMethodAsync>g__Awaited|12_0(ControllerActionInvoker invoker, ValueTask`1 actionResultValueTask)
   at Microsoft.AspNetCore.Mvc.Infrastructure.ControllerActionInvoker.<InvokeNextActionFilterAsync>g__Awaited|10_0(ControllerActionInvoker invoker, Task lastTask, State next, Scope scope, Object state, Boolean isCompleted)
   at Microsoft.AspNetCore.Mvc.Infrastructure.ControllerActionInvoker.Rethrow(ActionExecutedContextSealed context)
   at Microsoft.AspNetCore.Mvc.Infrastructure.ControllerActionInvoker.Next(State& next, Scope& scope, Object& state, Boolean& isCompleted)
   at Microsoft.AspNetCore.Mvc.Infrastructure.ControllerActionInvoker.InvokeInnerFilterAsync()
--- End of stack trace from previous location ---
   at Microsoft.AspNetCore.Mvc.Infrastructure.ResourceInvoker.<InvokeNextExceptionFilterAsync>g__Awaited|26_0(ResourceInvoker invoker, Task lastTask, State next, Scope scope, Object state, Boolean isCompleted)
   at Microsoft.AspNetCore.Mvc.Infrastructure.ResourceInvoker.Rethrow(ExceptionContextSealed context)
   at Microsoft.AspNetCore.Mvc.Infrastructure.ResourceInvoker.Next(State& next, Scope& scope, Object& state, Boolean& isCompleted)
   at Microsoft.AspNetCore.Mvc.Infrastructure.ResourceInvoker.InvokeNextResourceFilter()
--- End of stack trace from previous location ---
   at Microsoft.AspNetCore.Mvc.Infrastructure.ResourceInvoker.Rethrow(ResourceExecutedContextSealed context)
   at Microsoft.AspNetCore.Mvc.Infrastructure.ResourceInvoker.Next(State& next, Scope& scope, Object& state, Boolean& isCompleted)
   at Microsoft.AspNetCore.Mvc.Infrastructure.ResourceInvoker.InvokeFilterPipelineAsync()
--- End of stack trace from previous location ---
   at Microsoft.AspNetCore.Mvc.Infrastructure.ResourceInvoker.<InvokeAsync>g__Logged|17_1(ResourceInvoker invoker)
   at Microsoft.AspNetCore.Mvc.Infrastructure.ResourceInvoker.<InvokeAsync>g__Logged|17_1(ResourceInvoker invoker)
   at Microsoft.AspNetCore.Routing.EndpointMiddleware.<Invoke>g__AwaitRequestTask|6_0(Endpoint endpoint, Task requestTask, ILogger logger)
   at Microsoft.AspNetCore.ResponseCompression.ResponseCompressionMiddleware.InvokeCore(HttpContext context)
   at NSwag.AspNetCore.Middlewares.SwaggerUiIndexMiddleware.Invoke(HttpContext context)
   at NSwag.AspNetCore.Middlewares.RedirectToIndexMiddleware.Invoke(HttpContext context)
   at NSwag.AspNetCore.Middlewares.OpenApiDocumentMiddleware.Invoke(HttpContext context)
   at Microsoft.AspNetCore.Authorization.AuthorizationMiddleware.Invoke(HttpContext context)
   at Microsoft.AspNetCore.Authentication.AuthenticationMiddleware.Invoke(HttpContext context)
   at Microsoft.AspNetCore.Diagnostics.DeveloperExceptionPageMiddlewareImpl.Invoke(HttpContext context)
```

**Comment by adrianhall** [comment]
Understood.  I'm going to be working with the Cosmos SDK team on this topic.

**Comment by adrianhall** [comment]
I'm quickly becoming convinced that a generalized cosmos client is not possible because of the varying partition key requirements.  Whatever we do, we're not going to be able to derive the partition key from the ID in all cases.  If we do, we then are putting constraints on the ID beyond "it's a string", which breaks the "the client doesn't know what the server has implemented" thing.

We can certainly create, delete, update and use one of the fields within the object as the partition key(s), which then generates the ID.  The problem is that the ID is not generated by the server in the case of offline.  I'm not currently seeing a way around this as that ID is used as a key in the offline database.

Will need to do some more thinking about this, but it's non-obvious at this point in time.

**Comment by blackeyedyl** [comment]
I had a suspicion that this might be the outcome here and I agree with what you've said. The datasync library is designed to work with a single key to define a unit of data within a "table." Cosmos DB generally requires at least two keys per data unit—except in the specific case of using /id as the partition key. The single-key-per-unit approach aligns well with relational databases, which likely comprises the majority of the user base for this library. 

That said, Cosmos DB could be an excellent candidate for a fork of this library due to its built-in Version (_etag) and UpdatedAt (ts) for every document. With an updated API and using composite keys on the client SQLite database, a Cosmos-compatible version could be implemented with either EF Core or Cosmos SDK repositories.

For our purposes on this specific project, the v6 datasync library has enough benefits that we've decided to continue using it with the understanding that the cosmos native repository isn't going to be general enough for inclusion in the official library. Going forward I'll keep an eye on the v8 implementation to see if we want to go that direction: combining the offline/remote tables will be great for us to simplify some of our code, multi-item push will also be great.
